src/passes/pass.cpp

```diff
@@ -811,4 +810,25 @@ int PassRunner::getPassDebug() {
   return passDebug;
 }
 
+bool PassRunner::passRemovesDebugInfo(const std::string& name) {
+  return name == "strip" || name == "strip-debug" || name == "strip-dwarf";
+}
+
+bool PassRunner::shouldPreserveDWARF() {
+  // Check if the debugging subsystem wants to preserve DWARF.
+  if (!Debug::shouldPreserveDWARF(options, *wasm)) {
+    return false;
+  }
+
+  // We may need DWARF. Check if one of our previous passes would remove it
+  // anyhow, in which case, there is nothing to preserve.
+  for (auto& pass : passes) {
+    if (passRemovesDebugInfo(pass->name)) {
+      return false;
+    }
+  }
+
+  return true;
```

Comment:
These lines look like a duplicate of `willRemoveDebugInfo`,  worth merging?

Reply:
I'm not sure that's feasible. Especially after adding the caching, they look even less similar. One is to scan pass names 100% in advance, and the other is to check as we go while adding passes.